### ISSUE TITLE
Invalid VRRPv3 checksum & 2 masters in a 3 node cluster

### ISSUE THREAD

**Issue by vadapalliravikumar** [acassen/keepalived#1175]
In a 3 node cluster, invalid vrrpv3 checksum error is seen in 2 nodes, leading both the nodes to become masters. tcpdump showing bad checksum for vrrp packets coming from one of the node.

Using vrrp v3, keepalived unicast in a docker container environment.

Keepalived version:
```
root@maglev-master-1:/# keepalived -v
Keepalived v2.0.12 (01/25,2019), git commit v2.0.11-47-gce6cf748+

Copyright(C) 2001-2019 Alexandre Cassen, <acassen@gmail.com>

Built with kernel headers for Linux 4.4.170
Running on Linux 4.15.0-39-generic #42~16.04.1-Ubuntu SMP Wed Oct 24 17:09:54 UTC 2018

configure options: --prefix /usr --disable-dynamic-linking

Config options:  LVS VRRP VRRP_AUTH OLD_CHKSUM_COMPAT FIB_ROUTING

System options:  PIPE2 SIGNALFD INOTIFY_INIT1 VSYSLOG EPOLL_CREATE1 IPV4_DEVCONF IPV6_ADVANCED_API LIBNL3 RTA_ENCAP RTA_NEWDST RTA_PREF FRA_SUPPRESS_PREFIXLEN FRA_SUPPRESS_IFGROUP FRA_TUN_ID RTAX_CC_ALGO RTAX_QUICKACK RTEXT_FILTER_SKIP_STATS RTA_VIA FRA_OIFNAME IFA_FLAGS IP_MULTICAST_ALL LWTUNNEL_ENCAP_MPLS LWTUNNEL_ENCAP_ILA NET_LINUX_IF_H_COLLISION LIBIPTC_LINUX_NET_IF_H_COLLISION LIBIPVS_NETLINK IPVS_DEST_ATTR_ADDR_FAMILY IPVS_SYNCD_ATTRIBUTES IPVS_64BIT_STATS VRRP_VMAC CN_PROC SOCK_NONBLOCK SOCK_CLOEXEC O_PATH GLOB_BRACE INET6_ADDR_GEN_MODE VRF SO_MARK SCHED_RT SCHED_RESET_ON_FORK
root@maglev-master-1:/#
```

Config - Node A:
```
root@maglev-master-1:/# cat /etc/keepalived/keepalived.conf
global_defs {
  vrrp_version 3
  vrrp_iptables MAGLEV-KEEPALIVED-VIP
  enable_script_security
  script_user keepalived_script
  vrrp_garp_master_delay 40
  vrrp_garp_master_refresh 60
}

vrrp_script node_health_check {
  script       "/node_health_check.py"
  interval 60  # check every 60 seconds
  timeout 40   # Script Timeout of 40 seconds
  fall 3       # require 3 failures for FAULT Transition
}


vrrp_instance vip_10.195.76.227 {
  state BACKUP
  interface ens192
  virtual_router_id 88
  nopreempt
  advert_int 1

  track_interface {
    ens192
  }

  virtual_ipaddress {
    10.195.76.227 dev ens192
  }

  unicast_src_ip 10.195.76.224
  unicast_peer {
    10.195.76.225
    10.195.76.226
    }

  track_script {
    node_health_check
  }
}
```

Config - Node C:
```
root@maglev-master-10:/# cat /etc/keepalived/keepalived.conf
global_defs {
  vrrp_version 3
  vrrp_iptables MAGLEV-KEEPALIVED-VIP
  enable_script_security
  script_user keepalived_script
  vrrp_garp_master_delay 40
  vrrp_garp_master_refresh 60
}

vrrp_script node_health_check {
  script       "/node_health_check.py"
  interval 60  # check every 60 seconds
  timeout 40   # Script Timeout of 40 seconds
  fall 3       # require 3 failures for FAULT Transition
}


vrrp_instance vip_10.195.76.227 {
  state BACKUP 
  interface ens192
  virtual_router_id 88
  nopreempt
  advert_int 1

  track_interface {
    ens192
  }

  virtual_ipaddress {
    10.195.76.227 dev ens192
  }

  unicast_src_ip 10.195.76.226
  unicast_peer {
    10.195.76.224
    10.195.76.225
    }

  track_script {
    node_health_check
  }
}
root@maglev-master-10:/# 
```

Logs of Node A
```
Sat Mar 23 05:36:29 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:30 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:31 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:32 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:36 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:37 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:39 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:40 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:41 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:46 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:47 2019: Sending gratuitous ARP on ens192 for 10.195.76.227
Sat Mar 23 05:36:47 2019: (vip_10.195.76.227) Sending/queueing gratuitous ARPs on ens192 for 10.195.76.227
Sat Mar 23 05:36:47 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:48 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:49 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:50 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:51 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 05:36:55 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
```

logs of Node C
```
Sat Mar 23 02:57:31 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:57:32 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:57:33 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:57:34 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:57:35 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:57:36 2019: Sending gratuitous ARP on ens192 for 10.195.76.227
Sat Mar 23 02:57:36 2019: (vip_10.195.76.227) Sending/queueing gratuitous ARPs on ens192 for 10.195.76.227
Sat Mar 23 02:57:36 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:58:04 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:58:05 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:58:06 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:58:07 2019: (vip_10.195.76.227) Invalid VRRPv3 checksum
Sat Mar 23 02:58:08 2019: (vip_10.195.76.227) Invalid VRRPv3 
```
tcpdump of Node A:
```
$ sudo tcpdump -nl -vvvv -i any proto 112
tcpdump: listening on any, link-type LINUX_SLL (Linux cooked), capture size 262144 bytes
07:47:31.085512 IP (tos 0xc0, ttl 255, id 36912, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:31.085573 IP (tos 0xc0, ttl 255, id 21405, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.225: vrrp 10.195.76.224 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:31.085601 IP (tos 0xc0, ttl 255, id 21405, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.226: vrrp 10.195.76.224 > 10.195.76.226: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, (bad vrrp cksum 62d8), addrs: 10.195.76.227
07:47:32.085852 IP (tos 0xc0, ttl 255, id 21406, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.225: vrrp 10.195.76.224 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:32.085879 IP (tos 0xc0, ttl 255, id 21406, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.226: vrrp 10.195.76.224 > 10.195.76.226: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, (bad vrrp cksum 62d8), addrs: 10.195.76.227
07:47:32.085990 IP (tos 0xc0, ttl 255, id 36913, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:33.086235 IP (tos 0xc0, ttl 255, id 21407, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.225: vrrp 10.195.76.224 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:33.086276 IP (tos 0xc0, ttl 255, id 21407, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.226: vrrp 10.195.76.224 > 10.195.76.226: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, (bad vrrp cksum 62d8), addrs: 10.195.76.227
07:47:33.086338 IP (tos 0xc0, ttl 255, id 36914, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:34.086660 IP (tos 0xc0, ttl 255, id 21408, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.225: vrrp 10.195.76.224 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:34.086706 IP (tos 0xc0, ttl 255, id 36915, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:34.086735 IP (tos 0xc0, ttl 255, id 21408, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.226: vrrp 10.195.76.224 > 10.195.76.226: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, (bad vrrp cksum 62d8), addrs: 10.195.76.227
07:47:35.086974 IP (tos 0xc0, ttl 255, id 21409, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.225: vrrp 10.195.76.224 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:35.087017 IP (tos 0xc0, ttl 255, id 21409, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.226: vrrp 10.195.76.224 > 10.195.76.226: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, (bad vrrp cksum 62d8), addrs: 10.195.76.227
07:47:36.087105 IP (tos 0xc0, ttl 255, id 21410, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.225: vrrp 10.195.76.224 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:47:36.087142 IP (tos 0xc0, ttl 255, id 21410, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.226: vrrp 10.195.76.224 > 10.195.76.226: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, (bad vrrp cksum 62d8), addrs: 10.195.76.227
^C
16 packets captured
16 packets received by filter
0 packets dropped by kernel
```

tcpdump of Node C:
```
$ sudo tcpdump -nl -vv -i any proto 112
tcpdump: listening on any, link-type LINUX_SLL (Linux cooked), capture size 262144 bytes
07:46:22.500503 IP (tos 0xc0, ttl 255, id 36848, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:22.500549 IP (tos 0xc0, ttl 255, id 36848, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.225: vrrp 10.195.76.226 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:22.500909 IP (tos 0xc0, ttl 255, id 21380, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.226: vrrp 10.195.76.224 > 10.195.76.226: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, (bad vrrp cksum 62d8), addrs: 10.195.76.227
07:46:23.501284 IP (tos 0xc0, ttl 255, id 21381, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.224 > 10.195.76.226: vrrp 10.195.76.224 > 10.195.76.226: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, (bad vrrp cksum 62d8), addrs: 10.195.76.227
07:46:23.501293 IP (tos 0xc0, ttl 255, id 36849, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:23.501346 IP (tos 0xc0, ttl 255, id 36849, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.225: vrrp 10.195.76.226 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:24.501631 IP (tos 0xc0, ttl 255, id 36850, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:24.501668 IP (tos 0xc0, ttl 255, id 36850, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.225: vrrp 10.195.76.226 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:25.501742 IP (tos 0xc0, ttl 255, id 36851, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:25.501770 IP (tos 0xc0, ttl 255, id 36851, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.225: vrrp 10.195.76.226 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:26.501934 IP (tos 0xc0, ttl 255, id 36852, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:26.501988 IP (tos 0xc0, ttl 255, id 36852, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.225: vrrp 10.195.76.226 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:27.502200 IP (tos 0xc0, ttl 255, id 36853, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:27.502242 IP (tos 0xc0, ttl 255, id 36853, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.225: vrrp 10.195.76.226 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:28.502343 IP (tos 0xc0, ttl 255, id 36854, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.224: vrrp 10.195.76.226 > 10.195.76.224: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
07:46:28.502382 IP (tos 0xc0, ttl 255, id 36854, offset 0, flags [none], proto VRRP (112), length 32)
    10.195.76.226 > 10.195.76.225: vrrp 10.195.76.226 > 10.195.76.225: VRRPv3, Advertisement, vrid 88, prio 100, intvl 100cs, length 12, addrs: 10.195.76.227
^C
16 packets captured
39 packets received by filter
4 packets dropped by kernel
```

**Comment by vadapalliravikumar** [comment]
Another observation: restarting the keepalived container in Node A resolved the issue.

**Comment by pqarmitage** [comment]
This is very strange. Node A appears to have miscalculated the checksum for sending adverts to Node C (it should be 0x62d7, 0x62d8 is the checksum for adverts from Node A to Node B), and it also appears to be miscalculating the checksum that it should receive in adverts from Node C.

Is this problem reproducible, or has it occurred only once? If the problem is reproducible, could you please set out the steps to reproduce it. I have tried reproducing the problem and it doesn't produce the error for me.

Could it possibly be that there was a problem with the Node A container which caused the problem?

**Comment by vadapalliravikumar** [comment]
The issue is not seen every time, but we have seen it twice in less than a week.

The container image used for all the three nodes is same, and they are all spawned using the same orchestration (k8s). Are you suspecting anything specifically wrt the containers ?

**Comment by pqarmitage** [comment]
At the moment I am not sure what to suspect. No-one else has reported the problem, and I have certainly never seen it.

To help get an idea how frequently this is occurring, how many times have you restarted the containers/keepalived over the period when you have been experiencing the problem?

I don't think this will make any difference, but could you try upgrading your instances of keepalived to v2.0.14, just in case the problem has been resolved by a recent commit.

When the problem occurs, can you try making keepalived reload its configuration (send SIGHUP to the parent keepalived process). If that doesn't resolve the checksum problem, are you able to stop and restart keepalived without reloading the container?

**Comment by vadapalliravikumar** [comment]
@pqarmitage, could not re-create the issue in last few days. It was seen twice in a week and never after that.

As we don't know the exact steps to recreate the issue, not sure how it can be verified in v2.0.14.

We are keeping an eye on our daily regressions to see if this issue appears again. Meanwhile, can you see if the details given till now can be used to identify the root cause.

**Comment by pqarmitage** [comment]
@vadapalliravikumar Unfortunately I have no idea how this happened, and I can't think of anything in the code to look at. If it does occur again, could you try sending a SIGUSR1 signal to the parent keepalived process, and then post the /tmp/keepalived.data file, along with the keepalived configuration file. Doing this on all three nodes would be helpful

**Comment by pqarmitage** [comment]
Since this problem has only occurred twice, we cannot reproduce it and we cannot see what in the code could have caused this to happen, I'm closing this issue now. If it occurs again, then we can reopen the issue, and hope that further information is available that might identify the cause of the problem.

**Comment by vadapalliravikumar** [comment]
@pqarmitage, the issue is seen again. Collected the info that you have asked.

3 node cluster with IPs 10.106.172.121, 10.106.172.122, 10.106.172.123 & VIP = 10.106.172.124. keepalived unicast is used. 

Configs:
[121.keepalived.conf.txt](https://github.com/acassen/keepalived/files/3100479/121.keepalived.conf.txt)
[122.keepalived.conf.txt](https://github.com/acassen/keepalived/files/3100480/122.keepalived.conf.txt)
[123.keepalived.conf.txt](https://github.com/acassen/keepalived/files/3100481/123.keepalived.conf.txt)

Issue seen in the nodes 10.106.172.122 & 10.106.172.123. Sent SIGUSR1 and collected /tmp/keepalived.data from the 3 nodes in the cluster. 
keepalived.data:
[121.keepalived.data.txt](https://github.com/acassen/keepalived/files/3100482/121.keepalived.data.txt)
[122.keepalived.data.txt](https://github.com/acassen/keepalived/files/3100483/122.keepalived.data.txt)
[123.keepalived.data.txt](https://github.com/acassen/keepalived/files/3100484/123.keepalived.data.txt)

After that, I sent SIGHUP to the parent keepalived process in the node 10.106.172.122 & that resolved the issue. Below are the logs from 10.106.172.122 & 10.106.172.123 just before & after sending SIGHUP to the parent keepalived process in 10.106.172.122

10.106.172.122 logs:
```
RAVVADAP-M-K0RW:temp ravvadap$ cat 122.keepalived.logs 
Sat Apr 20 17:58:14 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:15 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:33 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:34 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:35 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:36 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:37 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:38 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:39 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:40 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:41 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:58:41 2019: (vip_10.106.172.124) Sending/queueing gratuitous ARPs on ens160 for 10.106.172.124
Sat Apr 20 17:58:41 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:54 2019: Reloading ...
Sat Apr 20 17:58:54 2019: Opening file '/etc/keepalived/keepalived.conf'.
Sat Apr 20 17:58:54 2019: Reloading
Sat Apr 20 17:58:54 2019: Opening file '/etc/keepalived/keepalived.conf'.
Sat Apr 20 17:58:54 2019: VRRP_Script(node_health_check) considered successful on reload
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) Ignoring track_interface ens160 since own interface
Sat Apr 20 17:58:54 2019: Assigned address 10.106.172.122 for interface ens160
Sat Apr 20 17:58:54 2019: Assigned address fe80::250:56ff:fead:a694 for interface ens160
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) ip address 10.106.172.124/32 dev ens160, no longer exist
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) setting VIPs.
Sat Apr 20 17:58:54 2019: VRRP sockpool: [ifindex(2), family(IPv4), proto(112), unicast(1), fd(10,11)]
Sat Apr 20 17:58:54 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) Sending/queueing gratuitous ARPs on ens160 for 10.106.172.124
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) Master received advert from 10.106.172.123 with same priority 100 but higher IP address than ours
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) Entering BACKUP STATE
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) removing VIPs.
```

10.106.172.123 logs:
```
RAVVADAP-M-K0RW:temp ravvadap$ cat 123.keepalived.logs 
Sat Apr 20 17:58:45 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:46 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:47 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:48 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:49 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:50 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:51 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:52 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:53 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) Invalid VRRPv3 checksum
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) Received advert from 10.106.172.122 with lower priority 100, ours 100, forcing new election
Sat Apr 20 17:58:54 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:58:54 2019: (vip_10.106.172.124) Sending/queueing gratuitous ARPs on ens160 for 10.106.172.124
Sat Apr 20 17:58:54 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:58:54 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:58:54 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:58:54 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:59:03 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:59:03 2019: (vip_10.106.172.124) Sending/queueing gratuitous ARPs on ens160 for 10.106.172.124
Sat Apr 20 17:59:34 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:59:34 2019: (vip_10.106.172.124) Sending/queueing gratuitous ARPs on ens160 for 10.106.172.124
Sat Apr 20 17:59:34 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
Sat Apr 20 17:59:34 2019: Sending gratuitous ARP on ens160 for 10.106.172.124
```

**Comment by vadapalliravikumar** [comment]
@pqarmitage, pls reopen the issue

**Comment by vadapalliravikumar** [comment]
@pqarmitage, were you able to identify the root cause ?

And, can you please suggest a workaround until the issue is closed.

**Comment by acassen** [comment]
will try to investigate deeper with Quentin, in the meantime would you please try fallback to VRRP_V2 ? this will trig another way to playground with csum. What is really strange in your case is, that it is perfectly working and this issue appear... looks really strange... Is there any chance something in the path can alter/recalculate csum ? using unicast in docker container...

I can't see log/conf for Node B.

have you tried to only use one unicast peer in unicast peer configuration block. You can create 2 VRRP instances, each one only with a single IP address in unicast_peer config block.

**Comment by vadapalliravikumar** [comment]
@acassen , configs of all 3 nodes are attached above. Logs of only 2 nodes, that are hitting checksum errors, are attached. The other node did not hit any checksum errors, so did not collect its logs.

We always run the cluster with 3 nodes (because of quorum requirements of etcd and other components). We never tried it with 2 nodes.

Shall wait for your analysis.

**Comment by acassen** [comment]
Ok, but could you try with vrrp_version 2

**Comment by vadapalliravikumar** [comment]
@acassen, given that the issue is not seen consistently with vrrp_version 3, I don't see how we can recreate & verify with vrrp_version 2.

We have been using vrrp_version 3 for a long time. And we cannot move to vrrp_version 2 at this point, as we dont want to expose ourselves to other unknown issues that might pop up with version 2.

Do you see any other way to identify the root cause ?

**Comment by pqarmitage** [comment]
We don't understand why this issue is happening, why it is so intermittent, or why no-one else is reporting it.

The only idea that I have had is if I write a patch to dump additional diagnostic information when the problem happens again. Would you be happy to build and run with a patched version in order to obtain more information about what is happening? If you can confirm you are happy to do this, and also that the version you are running is `Keepalived v2.0.12 (01/25,2019), git commit v2.0.11-47-gce6cf748+` then I can work on a patch for you.

**Comment by acassen** [comment]
unfortunately I didnt get a chance to inspect code deeper. But I did some initial investigations. According to the first msg in this thread. You did an unicast 3 nodes full-mesh. A(224) B(225) C(226). Reading carefully tcpdump output it appears A is causing issue. @pqarmitage I didnt see your second assertion about miscalculating csum from C. So postulate is 
(1) bad csum are coming from A
(2) bad csum are every time referring to the second unicast peer in the peer list
     => the corollary is adverts for first unicast peer are always with good csum

I would suspect that, maybe, there is something to look related to incremental csum that can/could potentially not be updated for second peer ? This was my initial conclusion before deep investigation and before reschedule to other matter.

If I am asking you to test with v2 is because it is a different way to compute csum and could offer you a way to workaround this issue.
Secondly I would strongly recommand using strong auth with unicast fashion conf. VRRP is a first hop segment routing election protocol, using unicast you are altering this postulate and open a way to unprotected injection (or you are enough confident with your networking isolation design).

**Comment by acassen** [comment]
hmm, maybe an issue with in_csum() comparing acc_sum. Will have a look at here tomorow.

**Comment by acassen** [comment]
The only way to remove from the equation any container issue playing in our back would be to add more debug as @pqarmitage suggested in order to track down csum sent and csum received. if they differ then something is altering in the path from node to node.

**Comment by vadapalliravikumar** [comment]
@acassen, @pqarmitage , the only way to re-create the issue is to commit the changes into our mainline and let it go through multiple regular regressions. One of those multiple regressions may see this issue, on one fine day. 

Given that it will go into the mainline, if the debugs are not too verbose and if they can be selectively enabled/disabled, we can take the patch.

**Comment by acassen** [comment]
this is a typical case where having a CLI could help live debugging. We could also use LOG_LEVEL to activate/inhibit live verbose but performing a global reload to take into account this updated config value will trig a global reload which apparently solve your issue.... hmm... interesting... maybe best solution would be to create a JSON listener to enable live interactions with Keepalived.

We really want to be sure Keepalived csum code is accurate, and IMHO having it running and producing good csum for long time and at one time trigging bad csum production leading to and infinite bad csum production is really strange. I would not suggest a cosmic ray hit your memory but I must admit idea cross my mind ;) but since it appears over time this certainly not arguing in favor of this explanation (except the fact we could be in a longterm solar wind)...

**Comment by acassen** [comment]
I read again the build and sending path... The code is good... the only option left would be to provide a quick patch with a 'slow-path' csum computation in oder to discriminate incremental csum update.

**Comment by pqarmitage** [comment]
I have added a patch to the checksum branch of my clone of keepalived at https://github.com/pqarmitage/keepalived/. The patch is applied to v2.0.12. The changes to the vanilla v2.0.12 are minimal, and other than adding some logging do not change it at all.

What the patch does is:
1. Log the checksum and packet contents the first time keepalived sends an advert to a destination.
2. Immediately before sending an advert it checks the checksum of the packet and logs if it is wrong. It also compares the checksum against the checksum sent in the previous packet to that destination and logs if it has changed.
3. When an advert is received it checks the checksum and if it is wrong, or if the checksum is different from the checksum last received it logs the details.

What the log entries look like are:
>Sun May  5 14:09:52 2019: (vip_10.195.76.227): First advert to 10.1.0.1, checksum: pkt 0x199c, calc 0x199c acc 0x2e661
Sun May  5 14:09:52 2019: IPv4 pseudo header - len 12
Sun May  5 14:09:52 2019: 0000  0a 01 00 03 0a 01 00 01  00 70 00 0c
Sun May  5 14:09:52 2019: Advert packet - len 32
Sun May  5 14:09:52 2019: 0000  45 c0 00 20 00 01 00 00  ff 70 00 00 0a 01 00 03
Sun May  5 14:09:52 2019: 0010  0a 01 00 01 31 58 c6 01  00 64 9c 19 0a c3 4c e3
Sun May  5 14:09:52 2019: (vip_10.195.76.227): First advert to 10.1.0.2, checksum: pkt 0x189c, calc 0x189c acc 0x2e761
Sun May  5 14:09:52 2019: IPv4 pseudo header - len 12
Sun May  5 14:09:52 2019: 0000  0a 01 00 03 0a 01 00 02  00 70 00 0c
Sun May  5 14:09:52 2019: Advert packet - len 32
Sun May  5 14:09:52 2019: 0000  45 c0 00 20 00 01 00 00  ff 70 00 00 0a 01 00 03
Sun May  5 14:09:52 2019: 0010  0a 01 00 02 31 58 c6 01  00 64 9c 18 0a c3 4c e3
Sun May  5 14:10:11 2019: vip_10.195.76.227: First received advert from 10.1.0.2, checksum: pkt 0x189b, calc 0x0000, acc 0x2fffd
Sun May  5 14:10:11 2019: IPv4 pseudo header - len 12
Sun May  5 14:10:11 2019: 0000  0a 01 00 02 0a 01 00 03  00 70 00 0c
Sun May  5 14:10:11 2019: Advert packet - len 32
Sun May  5 14:10:11 2019: 0000  45 c0 00 20 00 01 00 00  ff 70 a6 a6 0a 01 00 02
Sun May  5 14:10:11 2019: 0010  0a 01 00 03 31 58 c7 01  00 64 9b 18 0a c3 4c e3
Sun May  5 14:11:20 2019: (vip_10.195.76.227): Checksum valid change from 10.1.0.2 (priority 199 to 0), checksum: pkt 0x189b, previous 0x1962, calc 0x0000 acc 0x2fffd
Sun May  5 14:11:20 2019: IPv4 pseudo header - len 12
Sun May  5 14:11:20 2019: 0000  0a 01 00 02 0a 01 00 03  00 70 00 0c
Sun May  5 14:11:20 2019: Advert packet - len 32
Sun May  5 14:11:20 2019: 0000  45 c0 00 20 00 35 00 00  ff 70 a6 72 0a 01 00 02
Sun May  5 14:11:20 2019: 0010  0a 01 00 03 31 58 00 01  00 64 62 19 0a c3 4c e3

If there is a checksum error, the log message will contain `Checksum ERROR`.

You will need to build keepalived from the checksum branch referred to above, and run it on all three systems in the unicast group. Once the problem has occurred, we will need to see **all** keepalived logs from the time that keepalived started running on each system (the reason for this is that not only are the final log entries important, but also the initial log output, and any intervening output for checksum changes and also so that we can see if anything else is happening).

**Comment by acassen** [comment]
@vadapalliravikumar did you have a chance to use proposed debug release ?

**Comment by vadapalliravikumar** [comment]
@acassen, sorry missed your previous comment. Shall try the proposed debug release.

**Comment by acassen** [comment]
Hello @vadapalliravikumar, please let us know if you had a chance to run the referred release ? We would really like to look at debug infos provided when it start altering csum in order to fix it before next release. (I did my side an endurance test and just can't reproduce it)

**Comment by pqarmitage** [comment]
The branch created for the diagnostic patch for diagnosing the checksum problems referred to above, at [test branch with checksum diagnostics](https://github.com/acassen/keepalived/issues/1175#issuecomment-489427483) is being deleted due to no subsequent activity reported on running the diagnostic code. The patches for the diagnostics are attached [Patch file to apply to v2.0.12](https://github.com/acassen/keepalived/files/3327189/checksum-check.patch.txt) in case there is any further activity on this issue

The patch has now been forward ported to v2.0.17 and can be enabled with configuration option --enable-checksum-debug.

Since there has been no activity on this issue for nearly two months, the issue is now being closed. If there is any further information and feedback from the diagnostics, this issue can be reopened if necessary.
